runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_dotprod.c

```diff
@@ -75,3 +75,135 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0_1_2_4_8(
+  }
+
+  for (int k = 0; k < params->K; ++k) {
+    int8x16_t rhs[4];
+    for (int i = 0; i < 2; i++) {
+      int8x16_t r = vld1q_s8(rhs_ptr);
+      rhs_ptr += 16;
+      // We unpack int4s into individual int8s. To preserve signedness,
+      // int4s are moved to the upper 4-bits of each byte. This has the effect
+      // of multiplying each int4 by 2^4 = 16. To compensate, we divide the
+      // accumulator values by 16 before storing to memory.
+      // This int4 conversion trick is borrowed from the `qd8-f32-qc4w-gemm*`
+      // kernels in https://github.com/google/XNNPACK.
+      rhs[i + 0] = vshlq_n_s8(r, 4);
+      rhs[i + 2] = vandq_s8(r, vmask);
+    }
+
+    int8x8_t lhs[8];
+    if (M0 == 8) {
+      int8x16x2_t lhs_uzp_0 = vld2q_s8(lhs_ptr);
+      lhs_ptr += 32;
+      int8x16x2_t lhs_uzp_1 = vld2q_s8(lhs_ptr);
+      lhs_ptr += 32;
+      lhs[0] = vget_low_s8(lhs_uzp_0.val[0]);
+      lhs[1] = vget_low_s8(lhs_uzp_0.val[1]);
+      lhs[2] = vget_high_s8(lhs_uzp_0.val[0]);
+      lhs[3] = vget_high_s8(lhs_uzp_0.val[1]);
+      lhs[4] = vget_low_s8(lhs_uzp_1.val[0]);
+      lhs[5] = vget_high_s8(lhs_uzp_1.val[0]);
+      lhs[6] = vget_low_s8(lhs_uzp_1.val[1]);
+      lhs[7] = vget_high_s8(lhs_uzp_1.val[1]);
+    } else if (M0 == 4) {
+      int8x16x2_t lhs_uzp = vld2q_s8(lhs_ptr);
+      lhs_ptr += 32;
+      lhs[0] = vget_low_s8(lhs_uzp.val[0]);
+      lhs[1] = vget_low_s8(lhs_uzp.val[1]);
+      lhs[2] = vget_high_s8(lhs_uzp.val[0]);
+      lhs[3] = vget_high_s8(lhs_uzp.val[1]);
+    } else if (M0 == 2) {
+      int8x8x2_t lhs_uzp = vld2_s8(lhs_ptr);
+      lhs_ptr += 16;
+      lhs[0] = lhs_uzp.val[0];
+      lhs[1] = lhs_uzp.val[1];
+    } else {  // M0 == 1.
+      int8x8_t r = vld1_s8(lhs_ptr);
+      lhs_ptr += 8;
+      int8x8x2_t lhs_uzp = vuzp_s8(r, vzero);
+      lhs[0] = lhs_uzp.val[0];
+      lhs[1] = lhs_uzp.val[1];
+    }
+
+    // 1x2 * 2x8 -> 1x8.
+    acc[0] = vdotq_lane_s32(acc[0], rhs[0], lhs[0], 0);
+    acc[1] = vdotq_lane_s32(acc[1], rhs[1], lhs[0], 0);
+    acc[0] = vdotq_lane_s32(acc[0], rhs[2], lhs[1], 0);
+    acc[1] = vdotq_lane_s32(acc[1], rhs[3], lhs[1], 0);
+    if (M0 == 1) continue;
+    // 2x2 * 2x8 -> 2x8.
+    acc[2] = vdotq_lane_s32(acc[2], rhs[0], lhs[0], 1);
+    acc[3] = vdotq_lane_s32(acc[3], rhs[1], lhs[0], 1);
+    acc[2] = vdotq_lane_s32(acc[2], rhs[2], lhs[1], 1);
+    acc[3] = vdotq_lane_s32(acc[3], rhs[3], lhs[1], 1);
+    if (M0 == 2) continue;
+    // 4x2 * 2x8 -> 4x8.
+    acc[4] = vdotq_lane_s32(acc[4], rhs[0], lhs[2], 0);
+    acc[5] = vdotq_lane_s32(acc[5], rhs[1], lhs[2], 0);
+    acc[6] = vdotq_lane_s32(acc[6], rhs[0], lhs[2], 1);
+    acc[7] = vdotq_lane_s32(acc[7], rhs[1], lhs[2], 1);
+    acc[4] = vdotq_lane_s32(acc[4], rhs[2], lhs[3], 0);
+    acc[5] = vdotq_lane_s32(acc[5], rhs[3], lhs[3], 0);
+    acc[6] = vdotq_lane_s32(acc[6], rhs[2], lhs[3], 1);
+    acc[7] = vdotq_lane_s32(acc[7], rhs[3], lhs[3], 1);
+    if (M0 == 4) continue;
+    // 8x2 * 2x8 -> 8x8.
```

Comment:
The '2' dimension in these comments should be a '8'. It refers to the K0 value, since it is the dimension that is always the same in LHS and RHS.

Reply:
Thanks for catching that. Changed.